scripts/evolution.py

```diff
@@ -0,0 +1,460 @@
+import sys
+import os
+import space
+import numpy as np
+import csv
+import json
+import copy
+import datetime
+from jsonschema import Draft4Validator, validators, exceptions
+from utility_functions import *
+from collections import defaultdict
+from scipy import stats
+import numpy.random as rd
+import matplotlib
+# matplotlib.use('TkAgg')
+from matplotlib import pyplot as plt
+
+
+def get_best_config(configs):
+    """
+    Returns the configuration with lowest Value among the given configurations
+    :param configs: list of configurations.
+    :return: dict, the best configuration.
+    """
+    leader = configs[0]
+    for c in configs:
+        if c['Value'] < leader['Value']:
+            leader = c
+    return leader
+
+
+def mutation(param_space, config, mutationrate, list=False):
+    """
+    Mutates given configuration. This overcomplicates the procedure. But since I might
+    change the functionality I leave it like this for now
+    :param param_space: space.Space(), will give us information about parameters
+    :param configs: list of configurations.
+    :param mutationrate: integer for how many parameters to mutate
+    :param list: boolean whether returning one or more alternative configs
+    :return: list of dicts, list of mutated configurations
+    """
+
+    parameter_object_list = param_space.get_input_parameters_objects()
+    rd_config = dict()
+    for name, obj in parameter_object_list.items():
+        x = obj.randomly_select()
+        # sucks that there are parameters with only one value
+        if x == config[name]:
```

Comment:
Yes, the ideal scenario would be to sample until a new value is found. But in my nas problem  I have one ordinal parameter for the epochs the network should be trained for. This I gave a single value in my experiments, that is the problem. 

This should find a new value when possible and avoid getting stuck in loops, given I got the space.Space() commands right:

for name, obj in parameter_object_list.items():
        x = obj.randomly_select()
        single_valued_param = False
        param_type = param_space.get_type(obj)
        if param_type == 'real' or param_type == 'integer':
                if obj.get_max() == obj.get_min():
                        single_valued_param = True
        else:     # here ordinal and categorical parameters end up
               if obj.get_size() = 1:
                       single_valued_param = True

        while x == config[name] and single_valued_param == False:
                x = obj.randomly_select()
        rd_config[name] = x

Reply:
hmm, just realized that it will fail if there is a categorical/ordinal parameter with multiple values that all are the same. But why would anyone create such a parameter...?

Reply:
Your suggestion seems fine. I agree that users defining an input parameter with repeated values seems unlikely. If we want to be safe on those cases, we can do something similar to this:
https://github.com/luinardi/hypermapper/blob/d3318727dbce88eda56b533a089959eb106465ef/scripts/space.py#L1365

Where we halt if too many identical samples were made and warn the user.

Reply:
Do you think that would be better in this case?

Reply:
implemented both suggestions, now it should be quite robust.

https://github.com/samuellundberg/hypermapper/commit/0806fdc88ad9e9fb8f106a6e3aea81e9ba4c0221